### DISCUSSION TITLE
Override built-in parser config?

### DISCUSSION THREAD

**Discussion by helixbass** [nvim-treesitter/nvim-treesitter#4998] [Q&A] [answered]
**Is your feature request related to a problem? Please describe.**
What I'd like to do is to use a fork of the `tree-sitter-rust` Rust grammar

Specifically, I want to at least experiment with using a [version of the grammar](https://github.com/helixbass/tree-sitter-rust/pull/1) where macro invocation contents (eg `println!(...)`) are parsed "as Rust" (as opposed to "as token trees", which it looks like then maybe get re-parsed "as Rust" via the "injection" mechanism but only as part of syntax-highlighting ie `tree_sitter_highlight` if I'm understanding correctly?)

The reason I want to do this is that I've been writing some tree-sitter-based Lua code for doing editing actions and they don't work if the relevant part of the source file is inside a macro invocation (eg the contents of a `println!(...)` or a `vec![...]`)

**Describe the solution you'd like**
A way to "override" the built-in configuration for the Rust grammar (such that then it'll get automatically "picked up" and that grammar will get installed as expected etc) (this may already be possible I have no idea)

**Describe alternatives you've considered**
I tried manually deleting the `parser/rust.so` on the runtimepath and then re-launching neovim with this in my config:
```
local parser_config = require'nvim-treesitter.parsers'.get_parser_configs()
parser_config.rust = {
  install_info = {
    url = "https://github.com/helixbass/tree-sitter-rust",
    files = {'src/parser.c', 'src/scanner.c'},
    branch = 'macro-invocation-contents',
  },
}
require'nvim-treesitter.configs'.setup {
  ensure_installed = {"rust", "typescript", "tsx"}
}
```
but it seemed to not pick that up and instead just re-installed the "normal" Rust grammar

I'm guessing that for now I could just manually compile it and dump that `rust.so` into `parser/` and it would work, but would be nicer (at least in the longer term) for this to be something properly done via configuration so that eg it works on new machines

**Comment by clason** [comment]
You cannot do that without editing the lockfile as well. Either rename your parser, or -- as you write -- install it manually in your config directory (but don't forget the matching queries, too!)

**Comment by helixbass** [comment]
> Either rename your parser

Could you elaborate on this? Is there eg a way to specify that for the `rust` filetype I want to use the `rust_macro_invocation_contents` tree-sitter grammar?

> install it manually in your config directory

Ok I didn't see any explicit tree-sitter docs for how to build a `.so` for a given parser but so looking at the `nvim-treesitter` source I ran:
```
make --makefile ~/.local/share/nvim/site/pack/packer/start/nvim-treesitter/scripts/compile_parsers.makefile
```
in the local directory where I had the parser "built" (ie `src/parser.c` + `src/scanner.c` existed)
And that did create a `parser.so` file as expected

So then I _replaced_ the existing `rust.so` in `~/.local/share/nvim/site/pack/packer/start/nvim-treesitter/parser/` with that `parser.so` and restarted neovim but then it looks like neovim doesn't recognize that `rust.so` for some reason?

Specifically I get this error when trying to run `:InspectTree` with a Rust file open:
```
Error executing Lua callback: ...cos/share/nvim/runtime/lua/vim/treesitter/playground.lua:241: No parser available for the
 given buffer
stack traceback:
        [C]: in function 'assert'
        ...cos/share/nvim/runtime/lua/vim/treesitter/playground.lua:241: in function 'inspect_tree'
        ...tly/nvim-macos/share/nvim/runtime/lua/vim/treesitter.lua:506: in function 'inspect_tree'
        ...im-nightly/nvim-macos/share/nvim/runtime/plugin/nvim.lua:18: in function <...im-nightly/nvim-macos/share/nvim/r
untime/plugin/nvim.lua:9>
```

**Comment by amaanq** [comment]
rename parser.so to rust.so

**Reply by helixbass** [reply]
Right good guess but I did rename it:
```
$ ls -l ~/.local/share/nvim/site/pack/packer/start/nvim-treesitter/parser
total 3888
-rwxr-xr-x 1 jrosse staff  837240 Jun 23 13:17 rust.so*
-rwxr-xr-x 1 jrosse staff 1181184 Jun  7 16:47 tsx.so*
-rwxr-xr-x 1 jrosse staff 1148480 May 24 13:31 typescript.so*
```

**Reply by amaanq** [reply, marked as answer]
If you have the grammar project locally try this instead (it's what I do to use custom parsers)

```lua
local parser_configs = require("nvim-treesitter.parsers").get_parser_configs()
parser_configs["rust"] = {
  install_info = {
    url = "~/projects/treesitter/tree-sitter-rust",
    files = { "src/parser.c", "src/scanner.c" },
  },
}
```

**Reply by helixbass** [reply]
Whoa ok so ya using a local `url` like that (instead of a Github URL) does look like it successfully "overrode" the built-in Rust grammar config (after I deleted the existing `rust.so`)

That's great, thank you!

Also fwiw per the original reply above I additionally had to "override" the existing `queries/rust/*.scm` (I just copied the ones from `~/.local/share/nvim/site/pack/packer/start/nvim-treesitter/queries/rust/` into `~/.config/nvim/queries/rust/` and manually edited as needed) since the grammar change made some of those queries invalid and so neovim was "silently" choking on that (I had to run `:lua print(vim.inspect(vim.treesitter.get_parser(0, 'rust')))` to see a relevant error message)

**Reply by clason** [reply]
... or `:checkhealth`, or just opening the query in Neovim with the (new) parser in `runtimepath`.
